trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/TrinoLogsResource.java

```diff
@@ -0,0 +1,242 @@
+            default -> Response.status(BAD_REQUEST).build();
+        };
+    }
+
+    private Response getLogEvents(Request request)
+    {
+        byte[] bytes = request.requestContent().standardBytes().orElseThrow(() -> new WebApplicationException(BAD_REQUEST));
+        GetLogEventsRequest getLogEvents;
+        try {
+            getLogEvents = objectMapper.readValue(bytes, GetLogEventsRequest.class);
+        }
+        catch (IOException e) {
+            throw new WebApplicationException(e, BAD_REQUEST);
+        }
+
+        // TODO - support filter patterns
+
+        boolean startFromHead = getLogEvents.startFromHead.orElse(false);
+        Predicate<SaveEntry> predicate = entry -> startTimePasses(getLogEvents, entry) &&
+                endTimePasses(getLogEvents, entry) &&
+                logStreamNamesPasses(entry, getLogEvents);
+        int limit = getLogEvents.limit.orElse(Integer.MAX_VALUE);
+
+        List<SaveEntry> filteredEntries = loggerController.savedEntries(startFromHead, predicate);
+
+        record StartIndex(TokenType tokenType, int index) {}
+
+        // next token is merely an index
+        StartIndex startIndex = getLogEvents.nextToken().map(nextToken -> {
+            try {
+                TokenType tokenType;
+                if (nextToken.startsWith(TokenType.FORWARDS.prefix)) {
+                    tokenType = TokenType.FORWARDS;
+                }
+                else if (nextToken.startsWith(TokenType.BACKWARDS.prefix)) {
+                    tokenType = TokenType.BACKWARDS;
+                }
+                else {
+                    throw new WebApplicationException("Invalid nextToken", BAD_REQUEST);
+                }
+                return new StartIndex(tokenType, Integer.parseInt(nextToken.substring(2)));
+            }
+            catch (NumberFormatException _) {
+                throw new WebApplicationException("Invalid nextToken", BAD_REQUEST);
+            }
+        }).orElseGet(() -> new StartIndex(TokenType.FORWARDS, 0));
+
+        List<Event> events = filteredEntries.stream()
+                .skip(startIndex.index)
+                .limit(limit)
+                .map(entry -> new Event("trino", entry.entryId(), entry.timestamp().toEpochMilli(), entry.timestamp().toEpochMilli(), format(entry)))
+                .collect(toImmutableList());
+
+        Optional<String> nextBackwardToken;
+        Optional<String> nextForwardToken;
+        if (startIndex.tokenType == TokenType.FORWARDS) {
+            nextBackwardToken = TokenType.BACKWARDS.atIndex(startIndex.index);
+            nextForwardToken = ((startIndex.index + limit) < filteredEntries.size()) ? TokenType.FORWARDS.atIndex(startIndex.index + limit) : TokenType.FORWARDS.atIndex(startIndex.index);
+        }
+        else {
+            nextBackwardToken = ((startIndex.index - limit) >= 0) ? TokenType.BACKWARDS.atIndex(startIndex.index - limit) : TokenType.BACKWARDS.atIndex(startIndex.index);
```

Comment:
If we're going backwards, the offsets also start at 0 (since we've reversed the order of events when retrieving them). I realise now my previous comment wasn't very specific about this behaviour, sorry about that!

The tests appear to only be passing because the paginator always does forward pagination, even if the request is set to `startFromHead = false`.

Will commit a suggestion

Reply:
See https://github.com/vagaerg/s3-proxy/commit/c11bf5f93269a2d0d166f8b3de33599bcf134cab

That commit fixes the iteration order when traversing backwards and adds tests for all these cases.
I think longer term, we should use the requestNumber (which is monotonically increasing) to decide how to filter the logs, as the current approach (skipping entries in the log queue) will not be reliable

Reply:
Perhaps for a future PR, but we could use the `requestCounter` as part of the next/backwards token

We currently return something like `f/10` to mean "the next log entry is at index 10"

We could instead return `f/0:100:10` denoting "only consider log entries with request number 0 to 100, get the next entries starting at 10.

That way we get stable iteration, both forwards and backwards. It is possible entries get pruned from the list, but the meaning of a next/backward token never changes unlike the current behaviour

Reply:
Yes, I'd prefer to do that separately if needed